morango/syncsession.py

```diff
@@ -235,56 +244,175 @@ class SyncClient(object):
+            raise MorangoError('Chunk size must be evenly divisible by 100.')
+        self.chunk_size = chunk_size
+
+    def _starting(self, sync_filter, push):
+        data = None
+        # syncsession may or may not have created a transfer session
+        transfer_syncs = self.sync_session.transfersession_set.filter(filter=sync_filter).filter(active=True).filter(push=push)
+        if push:
+            if transfer_syncs:
+                logger.info("Resuming sync push...")
+                # grab active transfer session
+                self.current_transfer_session = transfer_syncs[0]
+                # turn off any other active transfer sessions attached to this syncsession
+                self.sync_session.transfersession_set.filter(active=True).exclude(id=self.current_transfer_session.id).update(active=False)
+                # clear buffered records from other transfer sessions
+                Buffer.objects.delete(transfersession_id__in=self.sync_session.transfersession_set.exclude(id=self.current_transfer_session.id).values_list('id', flat=True))
+            else:
+                logger.info('Beginning sync push...')
+                data = self._generate_transfer_session_data(True, sync_filter)
+                data.pop('last_activity_timestamp')
+                # create transfer session server side
+                try:
+                    response = self.sync_connection._create_transfer_session(data)
+                except requests.HTTPError as e:
+                    self.current_transfer_session.active = False
+                    self.current_transfer_session.save()
+                    raise
+
+                # create transfer session locally
+                data['server_fsic'] = response.json().get('server_fsic') or '{}'
+                data['last_activity_timestamp'] = timezone.now()
+                data['transfer_stage'] = transfer_status.QUEUING
+                self.current_transfer_session = TransferSession.objects.create(**data)
+        else:
+            if transfer_syncs:
+                logger.info("Resuming sync pull...")
+                self.current_transfer_session = transfer_syncs[0]
+                data = {
+                    'id': self.current_transfer_session.id,
+                    'filter': self.current_transfer_session.filter,
+                    'push': self.current_transfer_session.push,
+                    'sync_session_id': self.current_transfer_session.sync_session.id,
+                    'transfer_stage': transfer_status.QUEUING,
+                    'client_fsic': self.current_transfer_session.client_fsic,
+                }
+            else:
+                logger.info('Beginning sync pull...')
+                # create transfer session locally
+                data = self._generate_transfer_session_data(False, sync_filter)
+                data['last_activity_timestamp'] = timezone.now()
+                data['transfer_stage'] = transfer_status.QUEUING
```

Comment:
add assertions for client and server in same transfer stage
throwing custom exceptions